File: src/engine/ai/calculations/AiCalculations.java
package engine.ai.calculations;


import engine.ScoreBoard;
import engine.ai.calculations.stateCalculations.HillCalculations;
import engine.ai.calculations.stateCalculations.KillsCalculations;
import engine.ai.calculations.stateCalculations.RegicideCalculations;
import engine.ai.calculations.stateCalculations.TimedCalculations;
import engine.ai.controller.AiController;
import engine.entities.Player;
import engine.gameTypes.GameType;
import javafx.scene.shape.Rectangle;

public abstract class AiCalculations {
	
	//Game type object
	protected GameType gameType;
	//Score board object
	protected ScoreBoard scoreboard;
	//player object being controlled
	protected Player aiPlayer;
	//Map of game
	protected Rectangle map;
	//Players calculations object, used to carry out player related calculations
	protected PlayersCalculations playerCalc;
	//Movement calculations object, used to carry out movement related calculations
	protected MovementCalculations moveCalc;
	//Power up calculations object, used to carry out power up related calculations
	protected PowerupCalculations puCalc;
	//Time calculations object, used to carry out time related calculations
	protected TimeCalculations timeCalc;
	
	/**
	 * @param aiCon AI controller object
	 * @param map Map of game
	 * @param scoreboard score board object
	 * @param gameType Game type
	 */
	public AiCalculations(AiController aiCon,Rectangle map, ScoreBoard scoreboard, GameType gameType) {
		moveCalc = new MovementCalculations(aiCon, map);
		playerCalc = new PlayersCalculations(aiCon, map, scoreboard, gameType, moveCalc);
		puCalc = new PowerupCalculations(aiCon, map, playerCalc);
		timeCalc = new TimeCalculations();
		this.map = map;
		this.aiPlayer = aiCon.getAiPlayer();
	}
	
	/**
	 * initialises AI calculations object relevant to game type 
	 * @param map map of game
	 * @param scoreboard score board 
	 * @param gameType game type
	 */
	public static AiCalculations initializeAiCalculations(Rectangle map, ScoreBoard scoreboard, GameType gameType, AiController aiCon) {
		switch(gameType.getType()) {
		case FirstToXKills:
			return new KillsCalculations(aiCon, map, scoreboard, gameType);
		case Hill:
			return new HillCalculations(aiCon, map, scoreboard, gameType);
		case Regicide:
			return new RegicideCalculations(aiCon, map, scoreboard, gameType);
		case Timed:
			return new TimedCalculations(aiCon, map, scoreboard, gameType);
		default:
			return null;
		}
	}

	/**
	 * @return PlayersCalculations object
	 */
	public PlayersCalculations getPlayerCalc() {
		return playerCalc;
	}
	/**
	 * @return MovementCalculations object
	 */
	public MovementCalculations getMoveCalc() {
		return moveCalc;
	}
	
	/**
	 * @return PowerupCalculations object
	 */
	public PowerupCalculations getPowerupCalc() {
		return puCalc;
	}
	
	/**
	 * @return TimeCalculations object
	 */
	public TimeCalculations getTimeCalc() {
		return timeCalc;
	}
	
	
}

File: neon-element-cpp/src/engine/ai/calculations/AiCalculations.cpp
#include "AiCalculations.hpp"
#include "engine/ai/controller/AiController.hpp"
#include "engine/model/gametypes/HillGame.hpp"
#include "engine/model/gametypes/Regicide.hpp"
#include "engine/physics/CollisionDetector.hpp"

namespace neon {

AiCalculations::AiCalculations(AiController& aiCon, const Rect& map, ScoreBoard* scoreboard, GameType& gameType)
    : gameType_(gameType), scoreboard_(scoreboard), aiPlayer_(aiCon.getAiPlayer()),
      map_(map), moveCalc_(aiCon, map), playerCalc_(aiCon, map, scoreboard, moveCalc_),
      puCalc_(aiCon, map, playerCalc_)
{
}

std::unique_ptr<AiCalculations> AiCalculations::create(const Rect& map, ScoreBoard* scoreboard,
                                                        GameType& gameType, AiController& aiCon)
{
    switch (gameType.getType()) {
        case GameType::Type::FirstToXKills:
            return std::make_unique<KillsCalculations>(aiCon, map, scoreboard, gameType);
        case GameType::Type::Timed:
            return std::make_unique<TimedCalculations>(aiCon, map, scoreboard, gameType);
        case GameType::Type::Hill:
            return std::make_unique<HillCalculations>(aiCon, map, scoreboard, gameType);
        case GameType::Type::Regicide:
            return std::make_unique<RegicideCalculations>(aiCon, map, scoreboard, gameType);
    }
    return std::make_unique<KillsCalculations>(aiCon, map, scoreboard, gameType);
}

// --- HillCalculations ---

HillCalculations::HillCalculations(AiController& aiCon, const Rect& map, ScoreBoard* scoreboard, GameType& gameType)
    : AiCalculations(aiCon, map, scoreboard, gameType)
{
}

bool HillCalculations::onHill() const
{
    auto& hillGame = static_cast<HillGame&>(gameType_);
    Circle hill = hillGame.getHill();
    return CollisionDetector::checkCollision(hill, aiPlayer_->getBounds());
}

bool HillCalculations::closeToHill() const
{
    auto& hillGame = static_cast<HillGame&>(gameType_);
    Circle hill = hillGame.getHill();
    float dist = distance(aiPlayer_->getLocation(), Vec2{hill.x, hill.y});
    return dist < map_.width * 0.3f;
}

Player* HillCalculations::getOnHillPlayer()
{
    auto& hillGame = static_cast<HillGame&>(gameType_);
    Circle hill = hillGame.getHill();
    auto players = playerCalc_.getOtherPlayers();
    for (auto* p : players) {
        if (CollisionDetector::checkCollision(hill, p->getBounds()))
            return p;
    }
    return nullptr;
}

// --- RegicideCalculations ---

RegicideCalculations::RegicideCalculations(AiController& aiCon, const Rect& map, ScoreBoard* scoreboard, GameType& gameType)
    : AiCalculations(aiCon, map, scoreboard, gameType)
{
}

Player* RegicideCalculations::getKing()
{
    auto& regicide = static_cast<Regicide&>(gameType_);
    return regicide.getKing();
}

bool RegicideCalculations::kingIsClose()
{
    Player* king = getKing();
    if (!king || king == aiPlayer_) return false;
    return playerCalc_.isTooClose(king->getLocation());
}

} // namespace neon
